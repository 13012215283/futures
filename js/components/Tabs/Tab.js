import React from 'react';
import { Text, TouchableOpacity, ViewPropTypes } from 'react-native';
import PropTypes from 'prop-types';
import styles from './style';

const Tab = ({
  text,
  event,
  isActive,
  selectedTabStyle,
  selectedTextStyle,
  defaultTextStyle,
}) => (
  <TouchableOpacity
    style={[
      styles.iconView,
      isActive ? [styles.selectedWrapperStyle, selectedTabStyle] : {},
    ]}
    onPress={event}
  >
    <Text
      style={[
        styles.title,
        defaultTextStyle,
        isActive ? [styles.selectedTextStyle, selectedTextStyle] : {},
      ]}
    >
      {text}
    </Text>
  </TouchableOpacity>
);

export default Tab;

/* eslint-disable react/no-typos */
Tab.propTypes = {
  text: PropTypes.string.isRequired,
  event: PropTypes.func.isRequired,
  isActive: PropTypes.bool.isRequired,
  selectedTabStyle: ViewPropTypes.style,
  selectedTextStyle: Text.propTypes.style,
  defaultTextStyle: Text.propTypes.style,
};

Tab.defaultProps = {
  selectedTabStyle: {},
  selectedTextStyle: {},
  defaultTextStyle: {},
};
