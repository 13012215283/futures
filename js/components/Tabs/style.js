import { StyleSheet } from 'react-native';

import { sizes, colors } from '../themes';

const { f2 } = sizes;

export default StyleSheet.create({
  container: {
    position: 'relative',
    top: 0,
    right: 0,
    left: 0,
    height: 44,
    opacity: 1,
    backgroundColor: 'transparent',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  iconView: {
    flex: 1,
    height: 44,
    justifyContent: 'center',
    alignItems: 'center',
  },
  selectedWrapperStyle: {
    borderBottomWidth: 3,
    borderBottomColor: colors[1001],
  },
  selectedTextStyle: {
    color: colors[1001],
  },
  title: {
    fontSize: f2,
  },
});
