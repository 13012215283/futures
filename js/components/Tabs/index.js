import React, { Component } from 'react';
import {
  Text,
  View,
  PanResponder,
  Animated,
  ViewPropTypes,
} from 'react-native';
import PropTypes from 'prop-types';
import styles from './style';

import Tab from './Tab';

export default class Tabs extends Component {
  static propTypes = {
    tabNames: PropTypes.arrayOf(PropTypes.string).isRequired, // 包含选项卡名称的数组
    children: PropTypes.arrayOf(PropTypes.element).isRequired,
    containerStyle: ViewPropTypes.style,
    selectedTextStyle: Text.propTypes.style,
    selectedTabStyle: ViewPropTypes.style,
    defaultTextStyle: Text.propTypes.style,
    onChangeTab: PropTypes.func,
    activeIndex: PropTypes.number,
    swipeShield: PropTypes.number,
  };

  static defaultProps = {
    onChangeTab: () => {}, // 切换tab时调用的函数，返回当前tab的索引
    containerStyle: {}, // 外部样式
    selectedTextStyle: {}, // 选择某个tab的字体样式
    selectedTabStyle: {}, // 选择某个tab的样式
    defaultTextStyle: {}, // 默认的tab字体样式
    activeIndex: 0, // 指定激活的tab索引
    swipeShield: 100, // 滑动的阈值
  };

  constructor(props) {
    super(props);

    this.state = {
      activeIndex: this.validatedActiveIndex(this.props.activeIndex),
    };

    /* 开始滑动和结束滑动的横坐标x */
    this.startSwipe = 0;
    this.endSwipe = 0;

    this.onSelected = this.onSelected.bind(this);
  }

  componentWillMount() {
    if (this.props.swipeShield !== 0) {
      this.setPanResponder();
    }
  }

  /* 用户点击tab栏时触发 */
  onSelected(index) {
    this.setState({
      activeIndex: index,
    });
    this.props.onChangeTab(index);
  }

  getSwipedPageIndex(prevIdx, start, end) {
    const { swipeShield } = this.props;
    const moveVector = start - end;
    if (Math.abs(moveVector) < swipeShield) return prevIdx;
    return this.validatedActiveIndex(
      moveVector > 0 ? prevIdx - 1 : prevIdx + 1
    );
  }

  setPanResponder() {
    // 用来处理滑动手势的函数
    /* eslint-disable no-underscore-dangle */
    this._panResponder = PanResponder.create({
      // 要求成为响应者：
      onMoveShouldSetPanResponder: () => true,
      onMoveShouldSetPanResponderCapture: () => true,
      onPanResponderGrant: (evt, gestureState) => {
        // 开始手势操作。给用户一些视觉反馈，让他们知道发生了什么事情！
        // gestureState.{x,y}0 现在会被设置为0
        this.startSwipe = gestureState.moveX;
      },
      onPanResponderTerminationRequest: () => true,
      onPanResponderRelease: (evt, gestureState) => {
        // 用户放开了所有的触摸点，且此时视图已经成为了响应者。
        // 一般来说这意味着一个手势操作已经成功完成。
        this.endSwipe = gestureState.moveX;
        this.setState(prevState => {
          const prevActiveIdx = prevState.activeIndex;
          const changedIndex = this.getSwipedPageIndex(
            prevActiveIdx,
            this.startSwipe,
            this.endSwipe
          );

          if (prevActiveIdx !== changedIndex) {
            // 切换页时调用onChangeTab
            this.props.onChangeTab(changedIndex);
          }

          return { activeIndex: changedIndex };
        });
      },
    });
  }

  validatedActiveIndex(idx) {
    const { tabNames } = this.props;
    const len = tabNames.length;
    if (idx > len - 1) return len - 1;
    return idx < 0 ? 0 : idx;
  }

  render() {
    const translateX = Animated.add(1, 2).interpolate({
      inputRange: [-1, 0],
      outputRange: [-1, 0],
      extrapolate: 'clamp',
    });

    return (
      <View style={{ flex: 1 }}>
        <View style={[styles.container, this.props.containerStyle]}>
          {this.props.tabNames.map((text, index) => (
            <Tab
              key={text}
              text={text}
              event={() => this.onSelected(index)}
              isActive={index === this.state.activeIndex}
              selectedTabStyle={this.props.selectedTabStyle}
              selectedTextStyle={this.props.selectedTextStyle}
              defaultTextStyle={this.props.defaultTextStyle}
            />
          ))}
        </View>
        <Animated.View
          style={{ flex: 1, transform: [{ translateX }] }}
          {...(this.props.swipeShield !== 0
            ? this._panResponder.panHandlers
            : {})}
        >
          {this.props.children[this.state.activeIndex]}
        </Animated.View>
      </View>
    );
  }
}
