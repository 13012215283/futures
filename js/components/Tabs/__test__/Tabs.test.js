import { View, Text } from 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';
import Tabs from '../index';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Tabs tabNames={['item0', 'item1', 'item2']}>
        <View style={{ flex: 1 }}>
          <Text>Welcome to React Native!</Text>
        </View>
        <View style={{ flex: 1 }}>
          <Text>To get started, edit App.js</Text>
        </View>
        <View style={{ flex: 1 }}>
          <Text>xxxxxxx</Text>
        </View>
      </Tabs>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
