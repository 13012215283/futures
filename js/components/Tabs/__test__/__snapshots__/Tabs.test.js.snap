// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flexDirection": "row",
          "height": 44,
          "justifyContent": "center",
          "left": 0,
          "opacity": 1,
          "position": "relative",
          "right": 0,
          "top": 0,
        },
        Object {},
      ]
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "#e83b3b",
          "borderBottomWidth": 3,
          "flex": 1,
          "height": 44,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 28,
            },
            Object {},
            Array [
              Object {
                "color": "#e83b3b",
              },
              Object {},
            ],
          ]
        }
      >
        item0
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "height": 44,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 28,
            },
            Object {},
            Object {},
          ]
        }
      >
        item1
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "height": 44,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 28,
            },
            Object {},
            Object {},
          ]
        }
      >
        item2
      </Text>
    </View>
  </View>
  <View
    collapsable={undefined}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Welcome to React Native!
      </Text>
    </View>
  </View>
</View>
`;
